aten/src/ATen/native/cuda/Unique.cu

```diff
@@ -17,12 +22,100 @@ namespace native{
+  const bool consecutive,
+  const bool return_inverse,
+  const bool return_counts
+) {
+
+  cudaStream_t stream = at::cuda::getCurrentCUDAStream();
+
+  void *tmp_storage = nullptr;
+  size_t tmp_storage_size = 0;
+
+  int64_t num_inp = self.numel();
+  const scalar_t* self_data = self.data_ptr<scalar_t>();
+  int64_t *sorted_indices_data = nullptr;
+  Tensor sorted;
+  const scalar_t* sorted_data = self_data;
+
+  Tensor inverse_indices;
+
+  if (return_inverse) {
+    inverse_indices = at::empty(self.sizes(), self.options().dtype(kLong));
+  }
+
+  if (!consecutive) {
+    sorted = at::empty({num_inp}, self.options());
+    scalar_t *sorted_data_ = sorted.data_ptr<scalar_t>();
+    sorted_data = sorted_data_;  // const pointer
+    if (return_inverse) {
+      auto arange = at::arange(0, num_inp, self.options().dtype(kLong));
+      int64_t *arange_data = arange.data_ptr<int64_t>();
+      sorted_indices_data = static_cast<int64_t *>(c10::cuda::CUDACachingAllocator::raw_alloc(num_inp * sizeof(int64_t)));
+      // get tmp_storage_size
+      cub::DeviceRadixSort::SortPairs(tmp_storage, tmp_storage_size, self_data, sorted_data_, arange_data, sorted_indices_data, num_inp, 0, sizeof(scalar_t) * 8, stream);
+      // allocate tmp_storage and run the computation
+      tmp_storage = c10::cuda::CUDACachingAllocator::raw_alloc(tmp_storage_size);
+      cub::DeviceRadixSort::SortPairs(tmp_storage, tmp_storage_size, self_data, sorted_data_, arange_data, sorted_indices_data, num_inp, 0, sizeof(scalar_t) * 8, stream);
+    } else {
+      // get tmp_storage_size
+      cub::DeviceRadixSort::SortKeys(tmp_storage, tmp_storage_size, self_data, sorted_data_, num_inp, 0, sizeof(scalar_t) * 8, stream);
+      // llocate tmp_storage and run the computation
+      tmp_storage = c10::cuda::CUDACachingAllocator::raw_alloc(tmp_storage_size);
+      cub::DeviceRadixSort::SortKeys(tmp_storage, tmp_storage_size, self_data, sorted_data_, num_inp, 0, sizeof(scalar_t) * 8, stream);
+    }
+  }
+
+  // inverse indices
+  if (return_inverse) {
+    auto allocator = THCThrustAllocator(globalContext().lazyInitCUDA());
+    auto policy = thrust::cuda::par(allocator).on(stream);
+    int64_t* inverse_indices_ptr = inverse_indices.data_ptr<int64_t>();
+    Tensor inv_loc = consecutive ? inverse_indices.view(-1) : at::empty({num_inp}, self.options().dtype(kLong));
+    int64_t* inv_loc_ptr = inv_loc.data_ptr<int64_t>();
+    thrust::adjacent_difference(policy, sorted_data, sorted_data + num_inp, inv_loc_ptr, thrust::not_equal_to<scalar_t>());
+    inv_loc[0] = 0;
+    thrust::inclusive_scan(policy, inv_loc_ptr, inv_loc_ptr + num_inp, inv_loc_ptr);
+    if (!consecutive) {
+      thrust::scatter(policy, inv_loc_ptr, inv_loc_ptr + num_inp, sorted_indices_data, inverse_indices_ptr);
+      c10::cuda::CUDACachingAllocator::raw_delete(sorted_indices_data);
+    }
+  }
+
+  // cub always returns counts, so just ignore the return_counts flag and return counts always
```

Comment:
I'm not sure this is the right behavior. Allocate workspace for cub to put counts in, but don't return (defined tensor with) counts if they were not requested.

Reply:
I think it should be OK: It is never returned to the user. Because at python side, it is discarded by:

```python
    if return_inverse and return_counts:
        return output, inverse_indices, counts
    elif return_inverse:
        return output, inverse_indices
    elif return_counts:
        return output, counts
    else:
        return output
```

Reply:
I think this is OK: it is filtered out by python through
```python
    if return_inverse and return_counts:
        return output, inverse_indices, counts
    elif return_inverse:
        return output, inverse_indices
    elif return_counts:
        return output, counts
    else:
        return output
```